README.md
NaiveBayes
==========

**Naive Bayes text classification**

What is Naive bayes
-------------------

See also.

+ http://en.wikipedia.org/wiki/Naive_bayes


Tutorial
--------

The Bernoulli model.

+ http://nlp.stanford.edu/IR-book/html/htmledition/the-bernoulli-model-1.html

``` html
require 'naivebayes'
classifier = NaiveBayes::Classifier.new(:model => "berounoulli")
classifier.train("positive", {"aaa" => 0, "bbb" => 1})
classifier.train("negative", {"ccc" => 2, "ddd" => 3})
result = classifier.classify({"aaa" => 1, "bbb" => 1})
p result # => {"positive" => 0.8767123287671234,"negative" => 0.12328767123287669}
```

Relation to multinomial unigram language model.

+ http://nlp.stanford.edu/IR-book/html/htmledition/relation-to-multinomial-unigram-language-model-1.html

``` html
require 'naivebayes'
classifier = NaiveBayes::Classifier.new(:model => "multinomial")
classifier.train("positive", {"aaa" => 0, "bbb" => 1})
classifier.train("negative", {"ccc" => 2, "ddd" => 3})
result = classifier.classify({"aaa" => 1, "bbb" => 1})
p result # => {"positive" => 0.9411764705882353,"negative" => 0.05882352941176469}
```

Complement Naive Bayes.

+ http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.13.8572

``` html
require 'naivebayes'
classifier = NaiveBayes::Classifier.new(:model => "complement", :smoothing_parameter => 1)
classifier.train("positive", {"aaa" => 3, "bbb" => 1, "ccc" => 2})
classifier.train("negative", {"aaa" => 1, "bbb" => 4, "ccc" => 2})
classifier.train("neutral",  {"aaa" => 2, "bbb" => 3, "ccc" => 5})
result = classifier.classify({"aaa" => 4, "bbb" => 3, "ccc" => 3})
p result #=> {"neutral"=>9.985931139006835, "negative"=>10.112101263742268, "positive"=>10.836883752313222}
```


ChangeLog
---------

See doc/ChangeLog.


Developers
----------

See doc/AUTHORS.


Author
------

**774**

+ http://id774.net
+ http://github.com/id774


Copyright and license
---------------------

See the file doc/LICENSE.




doc/ChangeLog
=== 0.1.0 / 2014-01-23

* Implement Complement Naive Bayes.


=== 0.0.3 / 2013-12-13

* Accessable instance variable.

* Improvement specs.


=== 0.0.2 / 2013-07-09

* Append tutorial, demo script.


=== 0.0.1 / 2013-07-07

* First release.




spec/spec_helper.rb
# -*- coding: utf-8 -*-

APP_ROOT = File.expand_path(File.join(File.dirname(__FILE__), ".."))
$LOAD_PATH.unshift APP_ROOT
$LOAD_PATH.unshift File.join(APP_ROOT)
$LOAD_PATH.unshift File.join(APP_ROOT, 'lib')

require 'naivebayes'

if ENV['COVERAGE'] == 'on'
  require 'simplecov'
  require 'simplecov-rcov'
  SimpleCov.formatter = SimpleCov::Formatter::RcovFormatter

  SimpleCov.start do
    add_filter "spec"
    add_filter "vendor"
  end
end

unless /^1\.9\./ =~ RUBY_VERSION
  require 'rspec'
end

RSpec.configure do |config|
end


spec/lib/naivebayes_spec.rb
# -*- coding: utf-8 -*-

require File.dirname(__FILE__) + '/../spec_helper'

describe NaiveBayes do
  context "VERSION" do
    subject { NaiveBayes::VERSION }

    it { expect(subject).to eq "0.1.0-devel" }
  end
end


spec/lib/naivebayes/classifier_spec.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

require File.dirname(__FILE__) + '/../../spec_helper'

describe NaiveBayes::Classifier do
  describe '#initialize' do
    context '@frequency_table with berounoulli model' do
      subject { classifier.frequency_table }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return frequency table' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = {
          "positive" => {"aaa" => 2, "bbb" => 2},
          "negative" => {"ccc" => 2, "ddd" => 2}
        }

        expect(subject).to eq expected
      end
    end

    context '@frequency_table with multinomial model' do
      subject { classifier.frequency_table }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return frequency table' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = {
          "positive" => {"aaa" => 1, "bbb" => 3},
          "negative" => {"ccc" => 5, "ddd" => 7}
        }

        expect(subject).to eq expected
      end
    end

    context '@word_table with berounoulli model' do
      subject { classifier.word_table }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return word table' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = {
          "aaa" => 1, "bbb" => 1,
          "ccc" => 1, "ddd" => 1
        }

        expect(subject).to eq expected
      end
    end

    context '@word_table with multinomial model' do
      subject { classifier.word_table }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return word table' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = {
          "aaa" => 1, "bbb" => 1,
          "ccc" => 1, "ddd" => 1
        }

        expect(subject).to eq expected
      end
    end

    context '@instance_count_of with berounoulli model' do
      subject { classifier.instance_count_of }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return instance_count_of' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = {
          "positive" => 2,
          "negative" => 2
        }

        expect(subject).to eq expected
      end
    end

    context '@instance_count_of with multinomial model' do
      subject { classifier.instance_count_of }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return instance_count_of' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = {
          "positive" => 2,
          "negative" => 2
        }

        expect(subject).to eq expected
      end
    end

    context '@total_count with berounoulli model' do
      subject { classifier.total_count }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return total count' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = 4

        expect(subject).to eq expected
      end
    end

    context '@total_count with multinomial model' do
      subject { classifier.total_count }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return total count' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})
        classifier.train("positive", {"aaa" => 1, "bbb" => 2})
        classifier.train("negative", {"ccc" => 3, "ddd" => 4})

        expected = 4

        expect(subject).to eq expected
      end
    end

    context '@model with berounoulli model' do
      subject { classifier.model }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return model name' do
        expected = "berounoulli"
        expect(subject).to eq expected
      end
    end

    context '@model with multinomial model' do
      subject { classifier.model }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return model name' do
        expected = "multinomial"
        expect(subject).to eq expected
      end
    end

    context '@model with complement' do
      subject { classifier.model }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement") }

      it 'should return model name' do
        expected = "complement"
        expect(subject).to eq expected
      end
    end

    context '@smoothing_parameter with complement' do
      subject { classifier.smoothing_parameter }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement", :smoothing_parameter => 3) }

      it 'should return smoothing parameter' do
        expected = 3
        expect(subject).to eq expected
      end
    end
  end
end

describe NaiveBayes::Classifier do
  describe 'The berounoulli model' do
    context 'with train data of two expecting positive' do

      subject { classifier.classify({"aaa" => 1, "bbb" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "positive" => 0.8767123287671234,
          "negative" => 0.12328767123287669
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of two expecting negative' do

      subject { classifier.classify({"ccc" => 3, "ddd" => 3}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return negative' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "positive" => 0.12328767123287668,
          "negative" => 0.8767123287671234
        }

        expect(subject).to eq expected
      end
    end
  end
end

describe NaiveBayes::Classifier do
  describe 'The multinomial model' do
    context 'with train data of two expecting positive' do

      subject { classifier.classify({"aaa" => 1, "bbb" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "positive" => 0.9411764705882353,
          "negative" => 0.05882352941176469
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of two expecting negative' do

      subject { classifier.classify({"ccc" => 3, "ddd" => 3}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return negative' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "positive" => 0.0588235294117647,
          "negative" => 0.9411764705882353
        }

        expect(subject).to eq expected
      end
    end
  end
end

describe NaiveBayes::Classifier do
  describe 'Complement Naive Bayes' do
    context 'with train data of two expecting positive' do

      subject { classifier.classify({"aaa" => 1, "bbb" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement") }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "negative"=>0.8109302162163289,
          "positive"=>3.1986731175506815
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of two expecting negative' do

      subject { classifier.classify({"ccc" => 3, "ddd" => 3}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement") }

      it 'should return negative' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "positive"=>3.527593764407934,
          "negative"=>5.898526551448713
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of two expecting negative and smoothing parameter 1' do

      subject { classifier.classify({"ccc" => 3, "ddd" => 3}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement", :smoothing_parameter => 1) }

      it 'should return negative' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "positive"=>3.527593764407934,
          "negative"=>5.898526551448713
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of two expecting positive and smoothing parameter 3' do

      subject { classifier.classify({"aaa" => 1, "bbb" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement", :smoothing_parameter => 3) }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "negative"=>0.7137664677626813,
          "positive"=>1.9054187877005764
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of two expecting positive and smoothing parameter 5' do

      subject { classifier.classify({"aaa" => 1, "bbb" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement", :smoothing_parameter => 5) }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 0, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 3})

        expected = {
          "negative"=>0.7014459833746406,
          "positive"=>1.5040773967762742
        }

        expect(subject).to eq expected
      end
    end

  end
end

describe NaiveBayes::Classifier do
  describe 'The berounoulli model' do
    context 'with train data of three expecting positive' do

      subject { classifier.classify({"aaa" => 1, "bbb" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 2, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 2})
        classifier.train("neutral",  {"eee" => 3, "fff" => 3})

        expected = {
          "positive" => 0.7422680412371133,
          "negative" => 0.12886597938144329,
          "neutral"  => 0.12886597938144329
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of three expecting negative' do

      subject { classifier.classify({"ccc" => 3, "ddd" => 2}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return negative' do
        classifier.train("positive", {"aaa" => 2, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 2})
        classifier.train("neutral",  {"eee" => 3, "fff" => 3})

        expected = {
          "positive" => 0.12886597938144329,
          "negative" => 0.7422680412371133,
          "neutral"  => 0.12886597938144329
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of three expecting neutral' do

      subject { classifier.classify({"aaa" => 1, "ddd" => 2, "eee" => 3, "fff" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "berounoulli") }

      it 'should return neutral' do
        classifier.train("positive", {"aaa" => 2, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 2})
        classifier.train("neutral",  {"eee" => 3, "fff" => 3})

        expected = {
          "positive" => 0.2272727272727273,
          "negative" => 0.22727272727272724,
          "neutral"  => 0.5454545454545455
        }

        expect(subject).to eq expected
      end
    end
  end
end

describe NaiveBayes::Classifier do
  describe 'The multinomial model' do
    context 'with train data of three expecting positive' do

      subject { classifier.classify({"aaa" => 1, "bbb" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 2, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 2})
        classifier.train("neutral",  {"eee" => 3, "fff" => 3})

        expected = {
          "positive" => 0.896265560165975,
          "negative" => 0.06639004149377592,
          "neutral"  => 0.03734439834024896
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of three expecting negative' do

      subject { classifier.classify({"ccc" => 3, "ddd" => 2}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return negative' do
        classifier.train("positive", {"aaa" => 2, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 2})
        classifier.train("neutral",  {"eee" => 3, "fff" => 3})

        expected = {
          "positive" => 0.05665722379603399,
          "negative" => 0.9178470254957508,
          "neutral"  => 0.0254957507082153
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of three expecting neutral' do

      subject { classifier.classify({"aaa" => 1, "ddd" => 2, "eee" => 3, "fff" => 1}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "multinomial") }

      it 'should return neutral' do
        classifier.train("positive", {"aaa" => 2, "bbb" => 1})
        classifier.train("negative", {"ccc" => 2, "ddd" => 2})
        classifier.train("neutral",  {"eee" => 3, "fff" => 3})

        expected = {
          "positive" => 0.12195121951219513,
          "negative" => 0.09756097560975606,
          "neutral"  => 0.7804878048780488
        }

        expect(subject).to eq expected
      end
    end
  end
end

describe NaiveBayes::Classifier do
  describe 'Complement Naive Bayes' do
    context 'with train data of three expecting positive' do

      subject { classifier.classify({"aaa" => 4, "bbb" => 3, "ccc" => 3}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement") }

      it 'should return positive' do
        classifier.train("positive", {"aaa" => 3, "bbb" => 1, "ccc" => 2})
        classifier.train("negative", {"aaa" => 1, "bbb" => 4, "ccc" => 2})
        classifier.train("neutral",  {"aaa" => 2, "bbb" => 3, "ccc" => 5})

        expected = {
          "neutral"  => 9.985931139006835,
          "negative" => 10.112101263742268,
          "positive" => 10.836883752313222
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of three expecting negative' do

      subject { classifier.classify({"aaa" => 3, "bbb" => 4, "ccc" => 3}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement") }

      it 'should return negative' do
        classifier.train("positive", {"aaa" => 3, "bbb" => 1, "ccc" => 2})
        classifier.train("negative", {"aaa" => 1, "bbb" => 4, "ccc" => 2})
        classifier.train("neutral",  {"aaa" => 2, "bbb" => 3, "ccc" => 5})

        expected = {
          "neutral"  => 9.80360958221288,
          "positive" => 10.143736571753276,
          "negative" => 10.294422820536223
        }

        expect(subject).to eq expected
      end
    end

    context 'with train data of three expecting neutral' do

      subject { classifier.classify({"aaa" => 3, "bbb" => 3, "ccc" => 5}) }

      let(:classifier) { NaiveBayes::Classifier.new(:model => "complement") }

      it 'should return neutral' do
        classifier.train("positive", {"aaa" => 3, "bbb" => 1, "ccc" => 2})
        classifier.train("negative", {"aaa" => 1, "bbb" => 4, "ccc" => 2})
        classifier.train("neutral",  {"aaa" => 2, "bbb" => 3, "ccc" => 5})

        expected = {
          "negative" => 10.68941662877709,
          "positive" => 11.06002730362743,
          "neutral"  => 11.149081948812517
        }

        expect(subject).to eq expected
      end
    end
  end
end


lib/naivebayes/classifier.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

module NaiveBayes
  class Classifier
    attr_accessor :frequency_table, :word_table, :instance_count_of, :total_count, :model, :smoothing_parameter

    def initialize(params = {})
      @frequency_table = Hash.new
      @word_table = Hash.new
      @instance_count_of = Hash.new(0)
      @total_count = 0
      @model = params[:model]
      @smoothing_parameter = params[:smoothing_parameter] || 1
    end

    def train(label, feature)
      unless @frequency_table.has_key?(label)
        @frequency_table[label] = Hash.new(0)
      end
      feature.each {|word, frequency|
        if @model == "berounoulli"
          @frequency_table[label][word] += 1
        else
          @frequency_table[label][word] += frequency
        end
        @word_table[word] = 1
      }
      @instance_count_of[label] += 1
      @total_count += 1
    end

    def classify(feature)
      @model == "complement" ? cnb(feature) : mnb(feature)
    end

    private

    def mnb(feature)
      class_prior_of = Hash.new(1)
      likelihood_of = Hash.new(1)
      class_posterior_of = Hash.new(1)
      evidence = 0
      @instance_count_of.each {|label, freq|
        class_prior_of[label] = freq.to_f / @total_count.to_f
      }
      @frequency_table.each_key {|label|
        likelihood_of[label] = 1
        @word_table.each_key {|word|
          laplace_word_likelihood = (@frequency_table[label][word] + 1).to_f /
            (@instance_count_of[label] + @word_table.size()).to_f
          if feature.has_key?(word)
            likelihood_of[label] *= laplace_word_likelihood
          else
            likelihood_of[label] *= (1 - laplace_word_likelihood)
          end
        }
        class_posterior_of[label] = class_prior_of[label] * likelihood_of[label]
        evidence += class_posterior_of[label]
      }
      class_posterior_of.each {|label, posterior|
        class_posterior_of[label] = posterior / evidence
      }
      return class_posterior_of
    end

    def cnb(feature)
      all_class = @frequency_table.keys
      all_train_data = @instance_count_of.values.inject(0) {|s, v| s + v}
      class_posterior_of = all_class.map {|c|
        n_c = total_number_of_word_in_other_class(c)
        alpha = @smoothing_parameter*feature.length
        term2nd = feature.to_a.map {|e|
          k = e[0]
          v = e[1]
          v*Math.log((number_of_word_in_other_class(c, k) + @smoothing_parameter).to_f/(n_c + alpha))
        }.inject(0) {|s, v| s + v}
        theta_c = @instance_count_of[c].to_f/all_train_data
        [c, Math.log(theta_c) - term2nd]
      }.sort {|x, y| x[1] <=> y[1]}.flatten
      Hash[*class_posterior_of]
    end

    def total_number_of_word_in_other_class(c)
      all_words = @frequency_table.values.map {|h| h.keys}.flatten.sort.uniq
      other_classes = @frequency_table.keys - [c]
      other_classes.map {|c|
        all_words.map {|w|
          @frequency_table[c][w]
        }
      }.flatten.inject(0) {|s, v| s + v}
    end

    def number_of_word_in_other_class(c, i)
      other_classes = @frequency_table.keys - [c]
      other_classes.map {|c| @frequency_table[c][i]}.inject(0) {|s, v| s + v}
    end
  end
end


demo.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

$:.unshift File.join(File.dirname(__FILE__))

require 'lib/naivebayes'

puts "--- The Bernoulli model ---"
classifier = NaiveBayes::Classifier.new(:model => "berounoulli")

classifier.train("positive", {"aaa" => 0, "bbb" => 1})
classifier.train("negative", {"ccc" => 2, "ddd" => 3})
result = classifier.classify({"aaa" => 1, "bbb" => 1})

p result # => {"positive" => 0.8767123287671234,"negative" => 0.12328767123287669}

puts "--- Relation to multinomial unigram language model ---"

classifier = NaiveBayes::Classifier.new(:model => "multinomial")

classifier.train("positive", {"aaa" => 0, "bbb" => 1})
classifier.train("negative", {"ccc" => 2, "ddd" => 3})
result = classifier.classify({"aaa" => 1, "bbb" => 1})

p result # => {"positive" => 0.9411764705882353,"negative" => 0.05882352941176469}

puts "--- Complement Naive Bayes ---"

classifier = NaiveBayes::Classifier.new(:model => "complement", :smoothing_parameter => 1)

classifier.train("positive", {"aaa" => 3, "bbb" => 1, "ccc" => 2})
classifier.train("negative", {"aaa" => 1, "bbb" => 4, "ccc" => 2})
classifier.train("neutral",  {"aaa" => 2, "bbb" => 3, "ccc" => 5})
result = classifier.classify({"aaa" => 4, "bbb" => 3, "ccc" => 3})

p result #=> {"neutral"=>9.985931139006835, "negative"=>10.112101263742268, "positive"=>10.836883752313222}

